merge.go

```diff
@@ -0,0 +1,449 @@
+			delta.removeLabel = append(delta.removeLabel, db.GetLabel())
+		}
+		db.ForEachProperty(func(s string, i interface{}) bool {
+			_, ok := mem.GetProperty(s)
+			if !ok {
+				delta.removeProp = append(delta.removeProp, s)
+			}
+			return true
+		})
+	}
+	return delta
+}
+
+type OperationQueue struct {
+	ops []fmt.Stringer
+}
+
+type updateNode struct {
+	id     int64
+	labels []string
+	props  map[string]interface{}
+}
+
+type createNode struct {
+	id     int64
+	labels []string
+	props  map[string]interface{}
+}
+type createEdge struct {
+	id    int64
+	typ   []string
+	props map[string]interface{}
+}
+
+type updateEdge struct {
+	id    int64
+	typ   []string
+	props map[string]interface{}
+}
+
+func buildQueriesFromOperations(ops []fmt.Stringer) []string {
+	res := make([]string, 0)
+	for _, op := range ops {
+		res = append(res, op.String())
+	}
+	return res
+}
+
+func Merge(memGraph, dbGraph *lpg.Graph, dbGraphIds map[*lpg.Node]int64, dbEdges map[*lpg.Edge]int64) (*lpg.Graph, []string, error) {
+	qs := make([]string, 0)
+	memEntitiesMap := ls.GetEntityInfo(memGraph)
+	dbEntitiesMap := ls.GetEntityInfo(dbGraph)
+	findDBCounterpart := func(n *lpg.Node, schemaPV, idPV *ls.PropertyValue) *lpg.Node {
+		if ls.AsPropertyValue(n.GetProperty(ls.EntitySchemaTerm)).IsEqual(schemaPV) {
+			if ls.AsPropertyValue(n.GetProperty(ls.EntityIDTerm)).IsEqual(idPV) {
+				return n
+			}
+		}
+		return nil
+	}
+	memEntities := make([]*lpg.Node, 0, len(memEntitiesMap))
+	dbEntities := make([]*lpg.Node, 0, len(memEntitiesMap))
+	// align by index matching entity roots i.e, first entity in mem slice should match first entity in db slice
+	for n := range memEntitiesMap {
+		schemaPV := ls.AsPropertyValue(n.GetProperty(ls.EntitySchemaTerm))
+		eidPV := ls.AsPropertyValue(n.GetProperty(ls.EntityIDTerm))
+		memEntities = append(memEntities, n)
+		for db := range dbEntitiesMap {
+			if findDBCounterpart(db, schemaPV, eidPV) != nil {
+				dbEntities = append(dbEntities, db)
+				delete(dbEntitiesMap, db)
+			}
+		}
+	}
+	ops := make([]fmt.Stringer, 0)
+	for ix := 0; ix < len(memEntities); ix++ {
+		op, _ := mergeEntity(memEntities[ix], dbEntities[ix], dbGraphIds, dbEdges)
+		ops = append(ops, op...)
+	}
+	qs = append(qs, buildQueriesFromOperations(ops)...)
+	return dbGraph, qs, nil
+}
+
+func (dx delta) setDeltaToNode(dbNode *lpg.Node) {
+	for _, v := range dx.removeProp {
+		dbNode.RemoveProperty(v)
+	}
+	for k, v := range dx.setProp {
+		dbNode.SetProperty(k, v)
+	}
+	dbNode.SetLabels(lpg.NewStringSet(dx.addLabel...))
+}
+
+func (dx delta) setDeltaToEdge(dbEdge *lpg.Edge) {
+	for _, v := range dx.removeProp {
+		dbEdge.RemoveProperty(v)
+	}
+	for k, v := range dx.setProp {
+		dbEdge.SetProperty(k, v)
+	}
+	dbEdge.SetLabel(strings.Join(dx.addLabel, ","))
+}
+
+// type step struct {
+// 	edge *lpg.Edge
+// 	node *lpg.Node
+// }
+
+func matchDBChildNode(memItr, dbItr lpg.EdgeIterator, dbGraphIds map[*lpg.Node]int64, dbEdges map[*lpg.Edge]int64) []fmt.Stringer {
```

Comment:
Get memNode dbNode as args, not iterators. Iterators are single-use. Once they're done you cannot rewind them.

Don't return []stringer, return []delta